sklearn/model_selection/partition.py

```diff
@@ -0,0 +1,1108 @@
+    domain-specific stratification of the dataset.
+    """
+
+    def _iter_test_indices(self, y):
+        return range(self._sample_size(y))
+
+    def __repr__(self):
+        return '%s.%s(n=%i)' % (
+            self.__class__.__module__,
+            self.__class__.__name__,
+            self.n,
+        )
+
+    def __len__(self):
+        # TODO: remove?
+        return self.n
+
+
+class LeavePOut(_PartitionIterator):
+    """Leave-P-Out cross validation iterator
+
+    Provides train/test indices to split data in train test sets. This results
+    in testing on all distinct samples of size p, while the remaining n - p
+    samples form the training set in each iteration.
+
+    Note: ``LeavePOut(n, p)`` is NOT equivalent to ``KFold(n, n_folds=n // p)``
+    which creates non-overlapping test sets.
+
+    Due to the high number of iterations which grows combinatorically with the
+    number of samples this cross validation method can be very costly. For
+    large datasets one should favor KFold, StratifiedKFold or ShuffleSplit.
+
+    Parameters
+    ----------
+    n : int
+        Total number of elements in dataset.
+
+    p : int
+        Size of the test sets.
+
+    Examples
+    --------
+    >>> from sklearn import model_selection
+    >>> X = np.array([[1, 2], [3, 4], [5, 6], [7, 8]])
+    >>> y = np.array([1, 2, 3, 4])
+    >>> lpo = model_selection.LeavePOut(4, 2)
+    >>> len(lpo)
+    6
+    >>> print(lpo)
+    sklearn.model_selection.partition.LeavePOut(n=4, p=2)
+    >>> for train_index, test_index in lpo:
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    ...    X_train, X_test = X[train_index], X[test_index]
+    ...    y_train, y_test = y[train_index], y[test_index]
+    TRAIN: [2 3] TEST: [0 1]
+    TRAIN: [1 3] TEST: [0 2]
+    TRAIN: [1 2] TEST: [0 3]
+    TRAIN: [0 3] TEST: [1 2]
+    TRAIN: [0 2] TEST: [1 3]
+    TRAIN: [0 1] TEST: [2 3]
+    """
+
+    def __init__(self, n=None, p=None, indices=None):
+        super(LeavePOut, self).__init__(n, indices)
+        if p is None:
+            raise ValueError("LeavePOut: must supply p")
+        self.p = p
+
+    def _iter_test_indices(self, y):
+        for comb in combinations(range(self._sample_size(y)), self.p):
+            yield np.array(comb)
+
+    def __repr__(self):
+        return '%s.%s(n=%i, p=%i)' % (
+            self.__class__.__module__,
+            self.__class__.__name__,
+            self.n,
+            self.p,
+        )
+
+    def __len__(self):
+        # TODO: remove?
+        return int(factorial(self.n) / factorial(self.n - self.p)
+                   / factorial(self.p))
+
+
+class _BaseKFold(with_metaclass(ABCMeta, _PartitionIterator)):
+    """Base class to validate KFold approaches"""
+
+    @abstractmethod
+    def __init__(self, n, n_folds, indices, shuffle, random_state):
+        super(_BaseKFold, self).__init__(n, indices)
+
+        if abs(n_folds - int(n_folds)) >= np.finfo('f').eps:
```

Comment:
``type(n_folds)`` is not good, as it could be a numpy int. A ``isinstance(n_folds, numeric.Integer)`` would work, though.